plugins/autosuspend.koplugin/main.lua

```diff
@@ -43,90 +49,90 @@ function AutoSuspend:_enabled()
-                   self.settings_id)
-        return
-    end
 
-    local delay
+    local delay_suspend, delay_shutdown
 
     if PluginShare.pause_auto_suspend then
-        delay = self.auto_suspend_sec
+        delay_suspend = self.auto_suspend_sec
+        delay_shutdown = self.autoshutdown_sec
     else
-        delay = self.last_action_sec + self.auto_suspend_sec - os.time()
+        delay_suspend = self.last_action_sec + self.auto_suspend_sec - os.time()
+        delay_shutdown = self.last_action_sec + self.autoshutdown_sec - os.time()
     end
 
-    if delay <= 0 then
+    if delay_suspend <= 0 then
         logger.dbg("AutoSuspend: will suspend the device")
         UIManager:suspend()
+    elseif delay_shutdown <= 0 then
+        logger.dbg("AutoSuspend: initiating shutdown")
+        UIManager:poweroff_action()
     else
-        logger.dbg("AutoSuspend: schedule at ", os.time() + delay)
-        UIManager:scheduleIn(delay, function() self:_schedule(settings_id) end)
+        if self:_enabled() then
+            logger.dbg("AutoSuspend: schedule suspend at ", os.time() + delay_suspend)
+            UIManager:scheduleIn(delay_suspend, self._schedule, self)
+        end
+        if self:_enabledShutdown() then
+            logger.dbg("AutoSuspend: schedule shutdown at ", os.time() + delay_shutdown)
+            UIManager:scheduleIn(delay_shutdown, self._schedule, self)
+        end
     end
 end
 
-function AutoSuspend:_deprecateLastTask()
-    self.settings_id = self.settings_id + 1
-    logger.dbg("AutoSuspend: deprecateLastTask ", self.settings_id)
+function AutoSuspend:_unschedule()
+    logger.dbg("AutoSuspend: unschedule")
+    UIManager:unschedule(self._schedule)
 end
 
 function AutoSuspend:_start()
-    if self:_enabled() then
+    if self:_enabled() or self:_enabledShutdown() then
         logger.dbg("AutoSuspend: start at ", os.time())
         self.last_action_sec = os.time()
-        self:_schedule(self.settings_id)
+        self:_schedule()
     end
 end
 
 function AutoSuspend:init()
     UIManager.event_hook:registerWidget("InputEvent", self)
     self.auto_suspend_sec = self:_readTimeoutSec()
-    self:_deprecateLastTask()
+    self:_unschedule()
     self:_start()
+    -- self.ui is nil in the testsuite
+    if not self.ui or not self.ui.menu then return end
+    self.ui.menu:registerToMainMenu(self)
 end
 
 function AutoSuspend:onInputEvent()
     logger.dbg("AutoSuspend: onInputEvent")
     self.last_action_sec = os.time()
 end
 
--- We do not want auto suspend procedure to waste battery during suspend. So let's unschedule it
--- when suspending and restart it after resume.
 function AutoSuspend:onSuspend()
     logger.dbg("AutoSuspend: onSuspend")
-    self:_deprecateLastTask()
+    -- We do not want auto suspend procedure to waste battery during suspend. So let's unschedule it
+    -- when suspending and restart it after resume.
+    self:_unschedule()
+    if self:_enabledShutdown() then
```

Comment:
```suggestion
    if self:_enabledShutdown() and Device.wakeup_mgr then
```
Or something like that.